controllers/vspheredeploymentzone_controller_domain.go

```diff
@@ -0,0 +1,233 @@
+				ctx.Logger.V(4).Error(nil, "compute cluster does not own resource pool", "computeCluster", computeCluster, "resourcePool", ctx.VSphereDeploymentZone.Spec.PlacementConstraint.ResourcePool)
+				return errors.Errorf("compute cluster %s does not own resource pool %s", *computeCluster, resourcePool)
+			}
+		}
+	}
+	return nil
+}
+
+// verifyFailureDomain verifies the Failure Domain. It verifies the existence of tag and category specified and
+// checks whether the specified tags exist on the DataCenter or Compute Cluster or Hosts (in a HostGroup).
+func (r vsphereDeploymentZoneReconciler) verifyFailureDomain(ctx *context.VSphereDeploymentZoneContext, failureDomain infrav1.FailureDomain) error {
+	if _, err := ctx.AuthSession.TagManager.GetTagForCategory(ctx, failureDomain.Name, failureDomain.TagCategory); err != nil {
+		return errors.Wrapf(err, "failed to verify tag %s and category %s", failureDomain.Name, failureDomain.TagCategory)
+	}
+
+	managedRefFunc := getManagedRefFinder(failureDomain, ctx.VSphereFailureDomain.Spec.Topology, ctx.AuthSession.Finder)
+	managedRefs, err := managedRefFunc(ctx)
+	if err != nil {
+		return err
+	}
+
+TagVerifier:
+	for _, ref := range managedRefs {
+		attachedTags, err := ctx.AuthSession.TagManager.GetAttachedTags(ctx, ref)
+		if err != nil {
+			return err
+		}
+		for _, tag := range attachedTags {
+			if tag.Name == failureDomain.Name {
+				continue TagVerifier
+			}
+		}
+		return errors.Errorf("tag %s is not associated to %s", failureDomain.Name, ref.Reference().Type)
+	}
+	return nil
+}
+
+func (r vsphereDeploymentZoneReconciler) createAndAttachMetadata(ctx *context.VSphereDeploymentZoneContext, failureDomain infrav1.FailureDomain) error {
+	categoryID, err := metadata.CreateCategory(ctx, failureDomain.TagCategory, failureDomain.Type)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create category %s", failureDomain.TagCategory)
+	}
+	err = metadata.CreateTag(ctx, failureDomain.Name, categoryID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create tag %s", failureDomain.Name)
+	}
+
+	managedRefFunc := getManagedRefFinder(failureDomain, ctx.VSphereFailureDomain.Spec.Topology, ctx.AuthSession.Finder)
+	managedRefs, err := managedRefFunc(ctx)
+	if err != nil || len(managedRefs) == 0 {
+		return err
+	}
+
+	managedRef, err := managedRefFunc(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = ctx.AuthSession.TagManager.AttachTag(ctx, failureDomain.Name, managedRef[0])
+	if err != nil {
+		err = errors.Wrapf(err, "failed to attach tag %s to region", failureDomain.Name)
+	}
+	return err
+}
+
+// managedRefFinder is the method to find the reference of the type specified in the Failure Domain
+type managedRefFinder func(goctx.Context) ([]object.Reference, error)
+
+// getManagedRefFinder returns the method to find the reference specified in the Failure Domain
+func getManagedRefFinder(failureDomain infrav1.FailureDomain, topology infrav1.Topology, finder *find.Finder) managedRefFinder {
```

Comment:
this belongs into a govmomi package

Reply:
I specifically did not move it to the govmomi package since a lot of inputs are the CAPV API types. It isn't the case with the current setup of the `govmomi` package, but decided against it. 
On the other hand, I know this method has a lotta vSphere internals which do not need to reside in the controllers package. I will move it.